src/server/job_func.c

```diff
@@ -1205,6 +1205,13 @@ find_job(char *jobid)
 	if ((AVL_jctx != NULL) && ((pkey = svr_avlkey_create(buf)) != NULL)) {
 		if (avl_find_key(pkey, AVL_jctx) == AVL_IX_OK)
 			pj = (job *) pkey->recptr;
+		if (pj && (pj->ji_qs.ji_jobid[0] == 'X')) {
```

Comment:
find_job() seems like the wrong place to do this, shouldn't this be taken care of by job_free() or job_purge()?

Reply:
There has been an instance where a job already was found by `find_job()` during a user qdel request which lead to server coredump. Hence @alexis-cousein suggested to add this check to at least thwart a crash, while still looking for the root cause of stale job object in avl index. his analysis is pasted below verbatim


> GDB analysis:
User requests to delete a subjob of an array:
(gdb) print preq->rq_user
$5 = "lg4615", '\000' <repeats 250 times>
(gdb) print preq->rq_ind.rq_delete  
$1 = {rq_cmd = 1, rq_objtype = 2, rq_objname = "1492386[3].pbs", '\000' <repeats 259 times>, rq_attr = {ll_prior = 0x4043b1a8, ll_next = 0x4043b1a8,  
   ll_struct = 0x0}}
The job is "found" using AVL tree indices, but:
(gdb) print  &parent->ji_qs.ji_jobid                      
$16 = (char (*)[274]) 0x3364f968
(gdb) print  parent->ji_qs.ji_jobid   
$17 = "X492386[].pbs", '\000' <repeats 260 times>
(gdb) print  pjob->ji_qs.ji_jobid        
$18 = "X492386[3].pbs", '\000' <repeats 259 times>
The fact that the job ID starts with "X" means that we freed the corresponding structures -- this is used as a marker:

    /* if a subjob (of a Array Job), do not free certain items */
    /* which are malloced and shared with the parent Array Job */
    /* They will be freed when the parent is removed           */

    pj->ji_qs.ji_jobid[0] = 'X';    /* as a "freed" marker */
    free(pj);    /* now free the main structure */

> So I would bet that the job (and the parent) is nowhere to be found in svr_alljobs, but that the AVL tree indices were not updated correctly so we have a vestigial way to 'find' these when we should not. That lets users use "qdel" on subjobs purged from the history to crash the server.
I have not investigated the root cause of the fact the AVL tree still contains stale references, but in find_job adding:

     if ((AVL_jctx != NULL) && ((pkey = svr_avlkey_create(buf)) != NULL)) {
        if (avl_find_key(pkey, AVL_jctx) == AVL_IX_OK)
            pj = (job *) pkey->recptr;
        free(pkey);
        if ( pj != NULL && pj->ji_qs.ji_jobid[0] != 'X')
            return (pj);
        else
            /* AVL tree contains pointer to stale job */
            return NULL;
    }

> at least ensures that if the job structures haven't been overwritten we stand _some_ chance of "not finding" jobs that were freed in job_free.
It's not a complete fix, because if pj points to something already overwritten by heap reuse ji_qs.ji_jobid might point out of valid segments.

I have additionally added a call to `avl_delete_key()` to remove the stale record and also logged a error line, to leave an evidence in the log file

Reply:
Thanks for the explanation @Shrini-h . I was kind of suggesting the same thing: fix the bug that causes the free'd subjob to still be part of the job AVL tree, but do it inside job_free()/ job_purge(). The avl_delete_key() should be in job_free()/job_purge() right?

Reply:
> The avl_delete_key() should be in job_free()/job_purge() right?

Its in svr_dequejob() which is gets called by job_purge() not job_free() and based on what I know usage of job_free(), I don't think we can put AVL tree operation in job_free() as somewhere is hook code (IIRC, its in get_hook_results()) we make lite copy of job to add hook/job action after hook runs, and then we call job_free() on it, so if we put AVL operation in job_free() it will remove original job from AVL tree, and which will lead to many errors...

Reply:
Thank you @hirenvadalia for the insight.

@agrawalravi90 as @hirenvadalia mentioned there is already a code path from `job_purge()` that handles AVL record deletion for the job being purged, However since we have seen this instance the real unknown root cause is either making us to skip that code path while purging or the problem could be in the AVL tree operations itself, we might never know. So as stated this is just a guard and not a real bug fix.

Reply:
Since this is going into master (as opposed to a release branch), I think that we should spend the time to investigate the actual bug and fix that instead of adding a patch/guard which might mask the issue and possibly make it more difficult to debug later.

Also, If there was a bug in the AVL tree implementation, it would have shown up in other places as well, it's been there for years and is heavily used for almost every major operation done by the server, so I think we can probably rule that out and focus on the server's code path.

Reply:
+1 on @agrawalravi90's point, I don't think what this PR is doing is right fix, we should find real culprit and fix it, instead putting bandage on it.

Reply:
I too agree. I just hope that the original proposer of the patch (@alexis-cousein) also agrees on this. I will keep this PR open for some more time for further discussions on root causing

Reply:
Moreover we might not see this crash location again as it has been already removed previously in https://github.com/openpbs/openpbs/commit/7ce44ce2ada7e6333f32925c82845a620b66f0cd#diff-c3101b9054e0987316d8a49addb67e38L446